### DISCUSSION TITLE
How do I set up Redash server behind an nginx with path?

### DISCUSSION THREAD

**Discussion by eladitzhakian** [getredash/redash#6319] [Q&A] [answered]
Hey everyone

I want to set up a Redash server behind an nginx reverse-proxy, but I want it to sit behind a path, meaning:
```
location /redash {
   proxy_pass http://redash/;
}
```
Is there documentation regarding how to do this? There's usually an environment variable like `BASE_PATH` to help get around the path issues.

**Comment by aejazm** [comment]
I used [Caddy](https://caddyserver.com/) just to do this as I felt that things are more complex to configure in nginx. You just need the following configuration file for caddy to access redash using the URL https://yourhost.yourdomain.com

```
yourhost.yourdomain.com {
        log
        log {
           output file /var/log/caddy/access.log
        }
        reverse_proxy localhost:5000
        tls accountname@yourdomain.com
}
```

Assuming that redash is already setup to run on yourhost.yourdomain.com

Hope it helps.

**Reply by justinclift** [reply]
Note that I just reformatted that config section to be easier to read. :smile:

**Reply by eladitzhakian** [reply]
Thank you, however I need to use nginx with a path like `/redash`

**Comment by aejazm** [comment]
Thanks.
My first post....

**Reply by aejazm** [reply]
With nginx you could try the following. Note that I haven't tried this myself, so it could take some tweaking.

```
https{
...
server {
        listen       5000 default_server;
        listen       [::]:5000 default_server;
        server_name  _;
        
        location / {
             proxy_pass             http://localhost:5000;
             proxy_set_header       Host $host;
        }
  }
}

```

**Reply by eladitzhakian** [reply]
Thanks but let me reiterate that the path shouldnt be `/`, it should be `/redash` and thats why I believe further configuration is required

**Reply by justinclift** [reply]
As a rough guess there, making that `location` line say this *might* work:

```
location /redash {
```

**Reply by eladitzhakian** [reply]
@justinclift appreciate your suggestion but that won't work. you see redash's app has to be configured to use paths prefixed by `/redash`

**Reply by justinclift** [reply]
@eladitzhakian Yeah, now that you mention it this is sounding familiar.  Might be an environment variable or similar setting, to set the "slug".

The old online redash.io service used to have things named by "slug".  From rough memory:

* https://app.redash.io/customer1/
* https://app.redash.io/customer2/
* https://app.redash.io/customer3/

So, the "slug" thing would be the `customer1`, `customer2`, `customer3` in the examples above.  I'm saying "slug" because that's the term I remembering being used for that.

---

Maybe it's just the standard `REDASH_HOST` environment variable which needs changing?

https://redash.io/help/open-source/admin-guide/env-vars-settings

So, instead of `foo.example.com` it would be something like `foo.example.com/something`.

No clear idea (I'm very, very overdue for sleep), but it might be worth trying out. :smile:

**Comment by aejazm** [comment]
I didn't want /redash in the path, hence the need for tweaking...

**Comment by armando-fandango** [comment]
any solutions for this ?

**Comment by anemology** [comment, marked as answer]
Hope this can help, should work with SAML/LDAP login too. We are using SAML login and it's running for a while, no Redash code change needed.

Frontend use NGINX's `sub_filter` to replace `/` to `/redash`.
Backend use `SCRIPT_NAME` header let WSGI to prepend `/redash` when processing requests.

Redash version: `v10.1.0`, Docker Tag: `redash/redash:10.1.0.b50633`
NGINX version: `nginx:1.23.2-alpine`

```nginx
upstream redash_server {
    server 127.0.0.1:5000;
}

server {
    listen      443           ssl http2 default_server;
    listen [::]:443           ssl http2 default_server;

    access_log                /dev/stdout;
    error_log                 /dev/stderr info;

    include /etc/nginx/ssl.conf;

    location /redash/ {
        sub_filter ="/" ="/redash/";
        sub_filter ="/static/ ="/redash/static/;
        sub_filter ="/ldap/ ="/redash/ldap/;
        sub_filter ="/saml/login ="/redash/saml/login;
        sub_filter url(/static url(/redash/static;
        sub_filter_once off;
        sub_filter_types application/javascript text/css text/xml text/javascript application/json text/plain;

        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header SCRIPT_NAME /redash;

        proxy_pass       http://redash_server;
    }
}
```

Reference:

* [Hosting redash not under root](https://discuss.redash.io/t/hosting-redash-not-under-root/135)
* <https://wsgi.readthedocs.io/en/latest/definitions.html#envvar-SCRIPT_NAME>

**Reply by rajnathsah** [reply]
It works but while refreshing in url i can see location getting appended like this

redash/redash/redash/redash/redash/redash/redash/redash/redash

With each refresh, it is getting increased. Am i missing anything here?

**Reply by anemology** [reply]
@rajnathsah did you missing one of double quotes? Or you can post your nginx config and the URL encounter issue here.

**Reply by rajnathsah** [reply]
Here is my nginx conf

```
location /redash/ {
    sub_filter ="/" ="/redash/";
    sub_filter ="/static/ ="/redash/static/;
    sub_filter ="/ldap/ ="/redash/ldap/;
    sub_filter ="/saml/login ="/redash/saml/login;
    sub_filter url(/static url(/redash/static;
    sub_filter_once off;
    sub_filter_types application/javascript text/css text/xml text/javascript application/json text/plain;

    proxy_set_header Host $http_host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header SCRIPT_NAME /redash;

    proxy_pass  http://127.0.0.1:5000;
}
```

Let me know if you see anything wrong here?
